File: motor.cpp
#include "motor.hpp"
#include <unistd.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <iostream>

#define GPIO_BASE 0x3F200000  // Raspberry Pi 4 GPIO base address
#define GPIO_LEN  0xB4        // Length of the memory region

#define MOTOR_A_RPWM_PIN 12    // RPWM pin for Motor A
#define MOTOR_A_LPWM_PIN 13    // LPWM pin for Motor A
#define MOTOR_A_R_EN_PIN 16    // R_EN pin for Motor A
#define MOTOR_A_L_EN_PIN 19    // L_EN pin for Motor A

#define MOTOR_B_RPWM_PIN 20    // RPWM pin for Motor B
#define MOTOR_B_LPWM_PIN 21    // LPWM pin for Motor B
#define MOTOR_B_R_EN_PIN 26    // R_EN pin for Motor B
#define MOTOR_B_L_EN_PIN 6     // L_EN pin for Motor B

#define PWM_RANGE 100          // PWM range (0-100 for 100% duty cycle)

volatile unsigned int *gpio;

void setupGPIO() {
    int mem_fd = open("/dev/mem", O_RDWR | O_SYNC);
    if (mem_fd == -1) {
        std::cerr << "Failed to open /dev/mem!" << std::endl;
        exit(1);
    }

    gpio = (volatile unsigned int *)mmap(NULL, GPIO_LEN, PROT_READ | PROT_WRITE, MAP_SHARED, mem_fd, GPIO_BASE);
    if (gpio == MAP_FAILED) {
        std::cerr << "Memory mapping failed!" << std::endl;
        close(mem_fd);
        exit(1);
    }

    close(mem_fd);
}

void pinMode(int pin, int mode) {
    int reg = pin / 10;
    int shift = (pin % 10) * 3;
    if (mode == 1) {
        gpio[reg] |= (1 << shift);  // Set as output
    } else {
        gpio[reg] &= ~(1 << shift);  // Set as input
    }
}

void digitalWrite(int pin, int value) {
    int reg = (pin / 32) + 7;
    int shift = pin % 32;
    if (value == 1) {
        gpio[reg] = (1 << shift);  // Set pin high
    } else {
        gpio[reg] = (1 << (shift + 16));  // Set pin low
    }
}

void pwmWrite(int pin, int dutyCycle) {
    int pwmValue = (dutyCycle * PWM_RANGE) / 100;
    digitalWrite(pin, pwmValue);  // Simulating PWM control
}

void setupMotorPins() {
    setupGPIO();

    pinMode(MOTOR_A_RPWM_PIN, 1);
    pinMode(MOTOR_A_LPWM_PIN, 1);
    pinMode(MOTOR_A_R_EN_PIN, 1);
    pinMode(MOTOR_A_L_EN_PIN, 1);

    pinMode(MOTOR_B_RPWM_PIN, 1);
    pinMode(MOTOR_B_LPWM_PIN, 1);
    pinMode(MOTOR_B_R_EN_PIN, 1);
    pinMode(MOTOR_B_L_EN_PIN, 1);

    digitalWrite(MOTOR_A_R_EN_PIN, 1);
    digitalWrite(MOTOR_A_L_EN_PIN, 1);
    digitalWrite(MOTOR_B_R_EN_PIN, 1);
    digitalWrite(MOTOR_B_L_EN_PIN, 1);
}

void setMotorSpeed(int motor, int speed) {
    int pwm_duty_cycle = speed;

    if (motor == 1) {
        pwmWrite(MOTOR_A_RPWM_PIN, pwm_duty_cycle);
        pwmWrite(MOTOR_A_LPWM_PIN, pwm_duty_cycle);
    } else if (motor == 2) {
        pwmWrite(MOTOR_B_RPWM_PIN, pwm_duty_cycle);
        pwmWrite(MOTOR_B_LPWM_PIN, pwm_duty_cycle);
    }
}

void stopMotors() {
    digitalWrite(MOTOR_A_RPWM_PIN, 0);
    digitalWrite(MOTOR_A_LPWM_PIN, 0);
    digitalWrite(MOTOR_B_RPWM_PIN, 0);
    digitalWrite(MOTOR_B_LPWM_PIN, 0);
}

void forward() {
    // Both motors move forward at 50% duty cycle
    setMotorSpeed(1, 50); // Motor A
    setMotorSpeed(2, 50); // Motor B
    std::cout << "Moving forward at 50% duty cycle." << std::endl;
}

void reverse() {
    // Both motors move in reverse at 50% duty cycle
    setMotorSpeed(1, -50); // Motor A
    setMotorSpeed(2, -50); // Motor B
    std::cout << "Moving in reverse at 50% duty cycle." << std::endl;
}

void rightTurn() {
    // Left motor runs forward 40%, right motor runs reverse 10%
    setMotorSpeed(1, 40);  // Motor A (Left) Forward
    setMotorSpeed(2, -10); // Motor B (Right) Reverse
    std::cout << "Turning right." << std::endl;
}

void leftTurn() {
    // Left motor runs reverse 10%, right motor runs forward 40%
    setMotorSpeed(1, -10); // Motor A (Left) Reverse
    setMotorSpeed(2, 40);  // Motor B (Right) Forward
    std::cout << "Turning left." << std::endl;
}

void uTurn() {
    // Left motor runs reverse 40%, right motor runs forward 40%
    setMotorSpeed(1, -40); // Motor A (Left) Reverse
    setMotorSpeed(2, 40);  // Motor B (Right) Forward
    std::cout << "Making a U-turn." << std::endl;
}

int motor_init() {
    // Setup GPIO pins
    setupMotorPins();

    // Test different movements
  /*  forward();
    sleep(3); // Move forward for 3 seconds

    reverse();
    sleep(3); // Move reverse for 3 seconds

    rightTurn();
    sleep(3); // Turn right for 3 seconds

    leftTurn();
    sleep(3); // Turn left for 3 seconds

    uTurn();
    sleep(3); // U-turn for 3 seconds

    // Stop the motors after the movements
    stopMotors(); */
    std::cout << "Motors init completed" << std::endl;

    return 0;
}

File: motor.py
import RPi.GPIO as GPIO
import time

# Left driver pin setup
LEFT_RPWM = 12  # GPIO12 for Left motor forward
LEFT_LPWM = 13  # GPIO13 for Left motor reverse
LEFT_R_EN = 16  # GPIO16 for Left motor enable forward
LEFT_L_EN = 19  # GPIO19 for Left motor enable reverse

# Right driver pin setup
RIGHT_RPWM = 20  # GPIO20 for Right motor forward
RIGHT_LPWM = 21  # GPIO21 for Right motor reverse
RIGHT_R_EN = 26  # GPIO26 for Right motor enable forward
RIGHT_L_EN = 6   # GPIO6 for Right motor enable reverse

# PWM frequency (Hz)
PWM_FREQ = 500  # Recommended for L298N

# Global PWM instances
left_r_pwm = None
right_r_pwm = None
left_l_pwm = None
right_l_pwm = None

def motor_init():
    global left_r_pwm, right_r_pwm, left_l_pwm, right_l_pwm

    GPIO.setmode(GPIO.BCM)
    GPIO.setwarnings(False)

    # Set up pins
    GPIO.setup(LEFT_RPWM, GPIO.OUT)
    GPIO.setup(LEFT_LPWM, GPIO.OUT)
    GPIO.setup(LEFT_R_EN, GPIO.OUT)
    GPIO.setup(LEFT_L_EN, GPIO.OUT)

    GPIO.setup(RIGHT_RPWM, GPIO.OUT)
    GPIO.setup(RIGHT_LPWM, GPIO.OUT)
    GPIO.setup(RIGHT_R_EN, GPIO.OUT)
    GPIO.setup(RIGHT_L_EN, GPIO.OUT)

    # Enable motor driver
    GPIO.output(LEFT_R_EN, GPIO.HIGH)
    GPIO.output(LEFT_L_EN, GPIO.HIGH)
    GPIO.output(RIGHT_R_EN, GPIO.HIGH)
    GPIO.output(RIGHT_L_EN, GPIO.HIGH)

    # Initialize PWM for all pins
    left_r_pwm = GPIO.PWM(LEFT_RPWM, PWM_FREQ)   # Forward PWM for left motor
    right_r_pwm = GPIO.PWM(RIGHT_RPWM, PWM_FREQ) # Forward PWM for right motor
    left_l_pwm = GPIO.PWM(LEFT_LPWM, PWM_FREQ)   # Reverse PWM for left motor
    right_l_pwm = GPIO.PWM(RIGHT_LPWM, PWM_FREQ) # Reverse PWM for right motor

    # Start PWM with 0% duty cycle
    left_r_pwm.start(0)
    right_r_pwm.start(0)
    left_l_pwm.start(0)
    right_l_pwm.start(0)

    print("Motors initialized with PWM.")

def forward(duty_cycle=15):  # Reduced default duty cycle
    """
    Move forward at the specified duty cycle using PWM.
    """
    print(f"Moving forward at {duty_cycle}% duty cycle...")
    # Ensure reverse PWM is off
    left_l_pwm.ChangeDutyCycle(0)
    right_l_pwm.ChangeDutyCycle(0)
    # Set forward PWM to specified duty cycle
    left_r_pwm.ChangeDutyCycle(duty_cycle)
    right_r_pwm.ChangeDutyCycle(duty_cycle)

def reverse(duty_cycle=15):  # Reduced default duty cycle
    """
    Move reverse at the specified duty cycle using PWM.
    """
    print(f"Moving reverse at {duty_cycle}% duty cycle...")
    # Ensure forward PWM is off
    left_r_pwm.ChangeDutyCycle(0)
    right_r_pwm.ChangeDutyCycle(0)
    # Set reverse PWM to specified duty cycle
    left_l_pwm.ChangeDutyCycle(duty_cycle)
    right_l_pwm.ChangeDutyCycle(duty_cycle)

def turn_left(duty_cycle=15, duration=2):  # Reduced default duty cycle
    """
    Turn left: left motor reverse, right motor forward.
    """
    print(f"Turning left at {duty_cycle}% duty cycle...")
    left_r_pwm.ChangeDutyCycle(0)
    left_l_pwm.ChangeDutyCycle(duty_cycle)
    right_r_pwm.ChangeDutyCycle(duty_cycle)
    right_l_pwm.ChangeDutyCycle(0)
    time.sleep(duration)
    stop()

def turn_right(duty_cycle=15, duration=2):  # Reduced default duty cycle
    """
    Turn right: left motor forward, right motor reverse.
    """
    print(f"Turning right at {duty_cycle}% duty cycle...")
    left_r_pwm.ChangeDutyCycle(duty_cycle)
    left_l_pwm.ChangeDutyCycle(0)
    right_r_pwm.ChangeDutyCycle(0)
    right_l_pwm.ChangeDutyCycle(duty_cycle)
    time.sleep(duration)
    stop()

def u_turn(duty_cycle=15, duration=3):  # Reduced default duty cycle
    """
    Perform a U-turn: left motor forward, right motor reverse for a longer duration.
    """
    print(f"Performing U-turn at {duty_cycle}% duty cycle...")
    left_r_pwm.ChangeDutyCycle(duty_cycle)
    left_l_pwm.ChangeDutyCycle(0)
    right_r_pwm.ChangeDutyCycle(0)
    right_l_pwm.ChangeDutyCycle(duty_cycle)
    time.sleep(duration)
    stop()

def stop():
    """
    Stop the motors by setting the duty cycle to 0%.
    """
    print("Stopping motors...")
    left_r_pwm.ChangeDutyCycle(0)
    right_r_pwm.ChangeDutyCycle(0)
    left_l_pwm.ChangeDutyCycle(0)
    right_l_pwm.ChangeDutyCycle(0)

def cleanup():
    """
    Clean up GPIO and stop PWM.
    """
    print("Cleaning up GPIO and stopping PWM...")
    stop()
    left_r_pwm.stop()
    right_r_pwm.stop()
    left_l_pwm.stop()
    right_l_pwm.stop()
    GPIO.cleanup()
    print("GPIO cleaned up.")

if __name__ == "__main__":
    try:
        motor_init()
        forward(duty_cycle=15)  # Test at 15% duty cycle
        time.sleep(10)  # Move forward for 10 seconds
        stop()
    except KeyboardInterrupt:
        print("\nProgram interrupted by user.")
    except Exception as e:
        print(f"An error occurred: {e}")
    finally:
        cleanup()
